### DISCUSSION TITLE
Arduino UNO R4 WiFi freezing problem

### DISCUSSION THREAD

**Discussion by brsbnkc** [tdslite/tdslite#34] [Q&A]
I periodically write the data I collect with BME280 and DS3231 to MSSQL database, but sometimes it freezes while adding the data and the arduino gets stuck, how can we overcome this problem, my query for adding records is as follows;

``` int s1, s2, s3, s4;
  memset(SicaklikToleransi, 0, sizeof(SicaklikToleransi));
  SicaklikT.getText(SicaklikToleransi, sizeof(SicaklikToleransi));
  memset(NemToleransi, 0, sizeof(NemToleransi));
  NemT.getText(NemToleransi, sizeof(NemToleransi));
  memset(BasincToleransi, 0, sizeof(BasincToleransi));
  BasincT.getText(BasincToleransi, sizeof(BasincToleransi));
  if (isEmpty(SicaklikToleransi)) {
    strcpy(SicaklikToleransi, "0");
  }
  if (isEmpty(NemToleransi)) {
    strcpy(NemToleransi, "0");
  }
  if (isEmpty(BasincToleransi)) {
    strcpy(BasincToleransi, "0");
  }
  Sicaklik = atol(SicaklikToleransi) * 10;
  Nem = atol(NemToleransi) * 10;
  Basinc = atol(BasincToleransi) * 10;
  if (Sicaklik != 0) {
    s1 = (bme.readTemperature() * 10) + Sicaklik;
  } else {
    s1 = (bme.readTemperature() * 10);
  }
  if (Nem != 0) {
    s2 = (bme.readHumidity() * 10) + Nem;
  } else {
    s2 = (bme.readHumidity() * 10);
  }
  if (Basinc != 0) {
    s3 = (bme.readPressure() / 10) + Basinc;
    s4 = ((bme.readPressure() / 10) / 1.333) + Basinc;
  } else {
    s3 = (bme.readPressure() / 10);
    s4 = (bme.readPressure() / 10) / 1.333;
  }
  if (s1 != previousTemperature || s2 != previousHumidity || s3 != previousPressure || s4 != previousPressure2) {
    previousTemperature = s1;
    previousHumidity = s2;
    previousPressure = s3;
    previousPressure2 = s4;
    x0.setValue(s1);
    x1.setValue(s2);
    x2.setValue(s3);
    x3.setValue(s4);
    SensorAdiReadNexEeprom();
    if (isEmpty(Sensor)) {
      strcpy(Sensor, SensorCheck);
    } else {
      strcpy(SensorCheck, Sensor);
    }
    DateTime now = rtc.now();
    sprintf(query, "INSERT INTO AVS_ReadData ([Sicaklik], [Nem], [Basinc_hPa], [Basinc_mmHg], [SeriNo], [LokasyonAdi], [Tarih], [Saat]) VALUES (%d, %d, %d, %d, 'DC5475C3B7B4','%s','%d-%d-%d','%d:%d:%d');", s1, s2, s3, s4, Sensor, now.year(), now.month(), now.day(), now.hour(), now.minute(), now.second());
    //Serial.println(query);
    driver.execute_query(query);
  }```

**Comment by xmkg** [comment]
You are probably running out of memory, and I don't think this is related to tdslite itself in particular. If you have a minimal reproducer that only uses tdslite functions, I'd gladly investigate this issue. 

At this point, you have to check what's the real reason behind the freeze. First, I'd suggest checking the runtime memory usage as that is the most common problem behind the freezes.

Good luck.

**Reply by brsbnkc** [reply]
Thank you for your feedback.

When I remove the INSERT command in my code and do not add any records to the database, my code runs for days without freezing. Here I am using MS SQL express version and this server is running on a high specification server. The server and arduino are on the same subnet. And adding 7 different records in the same MSSQL server there is arduino. It freezes at different times on 7 devices. I also tried to operate these devices alone. There is still a freezing problem. No device other than Arduino has access to the server I specified. In this regard, is there a way to check whether a successful transaction is returned after the insert command?

**Reply by xmkg** [reply]
As I said previously, you should find out what is making your board freeze. Is it lack of memory, or an exception, or something else? Removing an INSERT clause hardly tell anything. 

You are moving a lot of memory around, so I'd start digging there.

Also, try the exact opposite of what you've tried, remove all your code and try to run a similar query periodically just using tdslite. If it still freezes, please let me know.

**Reply by brsbnkc** [reply]
Thank you. I'll do what you said and get back to it. Could it be a problem with the WiFi?

**Reply by xmkg** [reply]
Could be. There's no in-built reconnection mechanism in tdslite so if the connection between the client and the server is disrupted somehow, that could also appear as a "freeze".

**Reply by brsbnkc** [reply]
How can I disconnect when there is no response at a certain time after the mssql connection is established. Any chance to improve on this?

**Reply by brsbnkc** [reply]
By the way, I applied what you said, still the freezing problem persists while adding records. I constantly pinged the arduino from another device, the ping loss is 0, so there was no disconnection in wifi connection.

**Reply by brsbnkc** [reply]
Hello, the freezing problem is caused by your library. The reason why I came to this conclusion for sure is that I connected to MYSQL over the weekend and there was no freezing.

**Reply by xmkg** [reply]
Hi,

Can you post the full sketch you're using (the tdslite one)?

**Reply by brsbnkc** [reply]
If you share your e-mail address, I can send the entire code block there.

**Reply by xmkg** [reply]
hello@mkg.dev

**Reply by brsbnkc** [reply]
I sent.

**Comment by brsbnkc** [comment]
Hello,
I did what you said but the result is still the same. Is there anything else I should try? Or can I connect to the database before the INSERT command and close the connection after the record is successfully added?
